### PR TITLE
Fix sampling texture array types

### PR DESCRIPTION
**Pull request by magcius**

### CODE REVIEW

File: src/front/glsl/builtins.rs [hunk 1]
```diff
@@ -1670,12 +1670,19 @@ fn texture_call(
     meta: SourceMetadata,
 ) -> Result<Handle<Expression>> {
     if let Some(sampler) = ctx.samplers.get(&image).copied() {
+        let mut array_index = comps.array_index;
+
+        if let Some(mut array_index_expr) = array_index {
```
Comment (JCapucho) [inline src/front/glsl/builtins.rs:1675]: You can use `ref mut` here and remove the reference afterwards and the assignment

File: src/front/glsl/context.rs [hunk 2]
```diff
@@ -834,6 +834,25 @@ impl Context {
             .and_then(|(kind, width)| type_power(kind, width)))
     }
 
+    pub fn conversion(
```
Comment (JCapucho) [inline src/front/glsl/context.rs:837]: Very nice, this part really needed to refractored

**Review by JCapucho — CHANGES_REQUESTED** [review]
Did just light review since I won't be able to properly review the code until monday

File: src/front/glsl/builtins.rs (hunk 1 above)
Comment (magcius) [reply src/front/glsl/builtins.rs]: Thanks, that helps clean it up a lot.

File: src/front/glsl/types.rs [hunk 3]
```diff
@@ -131,14 +131,14 @@ pub fn parse_type(type_name: &str) -> Option<Type> {
 
                 let (dim, arrayed, class) = match size {
                     "1D" => (ImageDimension::D1, false, sampled(false)),
-                    "1DArray" => (ImageDimension::D1, false, sampled(false)),
+                    "1DArray" => (ImageDimension::D1, true, sampled(false)),
                     "2D" => (ImageDimension::D2, false, sampled(false)),
-                    "2DArray" => (ImageDimension::D2, false, sampled(false)),
+                    "2DArray" => (ImageDimension::D2, true, sampled(false)),
                     "2DMS" => (ImageDimension::D2, true, sampled(true)),
```
Comment (kvark) [inline src/front/glsl/types.rs:137]: why does this one have "true" for the arrayed?

File: tests/in/glsl/sampler2DArray.frag [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#version 440 core
```
Comment (kvark) [inline tests/in/glsl/sampler2DArray.frag:1]: I don't think we should have a separate test per sampler type. Can we just call it `samplers.frag`?

**Review by kvark — COMMENTED** [review]
Looks reasonable, a few notes here

File: src/front/glsl/types.rs (hunk 3 above)
Comment (magcius) [reply src/front/glsl/types.rs]: Nice catch!

File: tests/in/glsl/sampler2DArray.frag (hunk 4 above)
Comment (magcius) [reply tests/in/glsl/sampler2DArray.frag]: Done.

**Review by kvark — APPROVED** [review]

**Pull request merged**